File: english/data_processing/code/lidarp.m
% LiDAR data processing
 % load input data
 load lidar.txt;
 % data stored in array 'lidar' same as lidar = dlmread('lidar.txt', ',');
 % set window for processing
 xmin = 548040;
 ymin = 5129010;
 xmax = 548300;
 ymax = 5129270;
 dx = 10;    % GRID step
 dy = 10;
 % generate x & y grid
[x, y] = meshgrid(xmin:dx:xmax, ymin:dy:ymax);
 % grid interpolation for z
 z = griddata(lidar(:,1), lidar(:,2), lidar(:,3), x, y, 'linear');
 figure();    % 3D surface plot
 mesh(x, y, z);    % display mesh in 3D
 title('3D GRID');
 figure();
 meshc(x, y, z);   % display contours
 title('3D GRID and Contours');
 % contour lines
 figure(); contour(x, y, z);
 title('Contours');
 % volume calculation above 1000 m
 vol = 0;
 [r, c] = size(z);
 for ir = 1:r
     for ic = 1:c
         p = z(ir, ic);
         if (p > 1000)
             vol += dx * dy * (p - 1000);
         end
     end
 end
 printf('Volume above 1000m: %.0f m3\n', vol);
 % vectorized solution for volume calculation
 vol1 = sum(z(z > 1000) - 1000) * dx * dy;
 printf('Volume above 1000m: %.0f m3 (vectorized)\n', vol1);

File: english/data_processing/lessons/code/lidarp.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.interpolate import griddata
from mpl_toolkits.mplot3d import Axes3D

step = 10   # grid step
# read scattered points
a = np.loadtxt('lidar.txt', delimiter=',')
x = a[:,0]
y = a[:,1]
z = a[:,2]
# bounding box
xmin = 548040.0
ymin = 5129010.0
zmin = round(z.min() + 5.0, -1)
xmax = 548300.0
ymax = 5129270.0
zmax = round(z.max() - 5.0, -1)
# grid interpolation
xi = np.arange(xmin, xmax+step, step)
yi = np.arange(ymin, ymax+step, step)
xi, yi = np.meshgrid(xi, yi)
zi = griddata((x, y), z, (xi, yi), method='linear')
# 2D plot
fig = plt.figure()
plt.contour(xi,yi,zi,np.arange(zmin,zmax+step, step))
#plt.plot(x,y,'k.')
plt.xlabel('xi',fontsize=16)
plt.ylabel('yi',fontsize=16)
plt.savefig('lidar_contour.png',dpi=100)
plt.close(fig)
# 3D plot
fig = plt.figure()
ax = plt.axes(projection='3d')
ax.plot_wireframe(xi, yi, zi)
ax.contour3D(xi, yi, zi, 10)
ax.set_xlabel('x')
ax.set_ylabel('y')
ax.set_zlabel('z')
plt.savefig('lidar_3dcontour.png',dpi=100)
plt.close(fig)
# volume calculation
vol = sum(zi[zi > 1000] - 1000.0) * step**2
print('Volume above 1000m: {:.0f} m3'.format(vol))
